pkg/kvstore/etcd.go

```diff
@@ -0,0 +1,372 @@
+		if err != nil {
+			return 0, err
+		}
+		if value == nil {
+			// Something is really wrong
+			errMsg := "Unable to retrieve last free ID because the key is always empty\n"
+			log.Errorf(errMsg)
+			return 0, fmt.Errorf(errMsg)
+		}
+	}
+	var freeID uint32
+	if err := json.Unmarshal(value, &freeID); err != nil {
+		return 0, err
+	}
+	log.Debugf("Retrieving max free ID %d", freeID)
+	return freeID, nil
+}
+
+func (e *EtcdClient) SetMaxID(key string, firstID, maxID uint32) error {
+	value, err := e.GetValue(key)
+	if err != nil {
+		return err
+	}
+	if value == nil {
+		// FreeID is empty? We should set it out!
+		log.Infof("Empty FreeID, setting it up with default value %d", firstID)
+		if err := e.InitializeFreeID(key, firstID); err != nil {
+			return err
+		}
+		k, err := e.GetValue(key)
+		if err != nil {
+			return err
+		}
+		if k == nil {
+			// Something is really wrong
+			errMsg := "Unable to setting ID because the key is always empty\n"
+			log.Errorf(errMsg)
+			return fmt.Errorf(errMsg)
+		}
+	}
+	return e.SetValue(key, maxID)
+}
+
+func (e *EtcdClient) GASNewSecLabelID(basePath string, baseID uint32, secCtxLabels *types.SecCtxLabel) error {
+	setID2Label := func(lockPair KVLocker) error {
+		defer lockPair.Unlock()
+		secCtxLabels.ID = baseID
+		keyPath := path.Join(basePath, strconv.FormatUint(uint64(secCtxLabels.ID), 10))
+		if err := e.SetValue(keyPath, secCtxLabels); err != nil {
+			return err
+		}
+		return e.setMaxLabelID(baseID + 1)
+	}
+
+	beginning := baseID
+	for {
+		log.Debugf("Trying to acquire a new free ID %d", baseID)
+		keyPath := path.Join(basePath, strconv.FormatUint(uint64(baseID), 10))
+
+		locker, err := e.LockPath(GetLockPath(keyPath))
+		if err != nil {
+			return err
+		}
+
+		value, err := e.GetValue(keyPath)
+		if err != nil {
+			locker.Unlock()
+			return err
+		}
+		if value == nil {
+			return setID2Label(locker)
+		}
+		var consulLabels types.SecCtxLabel
+		if err := json.Unmarshal(value, &consulLabels); err != nil {
+			locker.Unlock()
+			return err
+		}
+		if consulLabels.RefCount() == 0 {
+			log.Infof("Recycling ID %d", baseID)
+			return setID2Label(locker)
+		}
+		locker.Unlock()
+
+		baseID++
+		if baseID > common.MaxSetOfLabels {
+			baseID = common.FirstFreeLabelID
+		}
+		if beginning == baseID {
+			return fmt.Errorf("reached maximum set of labels available.")
+		}
+	}
+}
+
+func (e *EtcdClient) updateSecLabelIDRef(secCtxLabels types.SecCtxLabel) error {
+	key := path.Join(common.LabelIDKeyPath, strconv.FormatUint(uint64(secCtxLabels.ID), 10))
+	return e.SetValue(key, secCtxLabels)
+}
+
+func (e *EtcdClient) setMaxLabelID(maxID uint32) error {
+	return e.SetMaxID(common.LastFreeLabelIDKeyPath, common.FirstFreeLabelID, maxID)
+}
+
+func (e *EtcdClient) setMaxServiceL4ID(maxID uint32) error {
+	return e.SetMaxID(common.LastFreeServiceIDKeyPath, common.FirstFreeServiceID, maxID)
+}
+
+func (e *EtcdClient) GASNewServiceL4ID(basePath string, baseID uint32, sl4 *types.ServiceL4ID) error {
```

Comment:
OK. Could we then document these methods to make it easy for the reader to understand their differences?